### PR TITLE
Test create custom layer in python

### PR DESCRIPTION
**Pull request by l-bat**
```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2019r3.0:16.04
build_image:Custom Win=openvino-2019r3.0
build_image:Custom Mac=openvino-2019r3.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/dnn/misc/python/test/test_dnn.py [hunk 1]
```diff
@@ -226,6 +226,67 @@ def test_async(self):
...
+        net = cv.dnn.readNetFromCaffe(str.encode(proto))
+        src_shape = [1, 2, 5, 5]
+        dst_shape = [1, 2, 3, 3]
+        inp = np.arange(0, np.prod(src_shape)).reshape(src_shape)
+        roi = np.empty(dst_shape)
+        net.setInput(inp, "input")
+        net.setInput(roi, "roi")
+        out = net.forward()
+        ref = inp[:, :, 1:4, 1:4]
+        assert(np.allclose(out, ref))
```
Comment (dkurt) [inline modules/dnn/misc/python/test/test_dnn.py]: Please use `normAssert`

File: modules/dnn/misc/python/test/test_dnn.py [hunk 2]
```diff
@@ -226,6 +226,67 @@ def test_async(self):
...
+              bottom: "input"
+              bottom: "roi"
+              top: "Crop"
+            }'''
+
+        net = cv.dnn.readNetFromCaffe(str.encode(proto))
+        src_shape = [1, 2, 5, 5]
+        dst_shape = [1, 2, 3, 3]
+        inp = np.arange(0, np.prod(src_shape)).reshape(src_shape)
+        roi = np.empty(dst_shape)
```
Comment (dkurt) [inline modules/dnn/misc/python/test/test_dnn.py]: Please check that input data is `float32`.

File: modules/python/src2/cv2.cpp [hunk 3]
```diff
@@ -392,6 +392,7 @@ static bool pyopencv_to(PyObject* o, Mat& m, const ArgInfo& info)
             oarr = (PyArrayObject*) o;
         }
         else {
+            CV_Assert(PyArray_ISCONTIGUOUS(oarr));
             oarr = PyArray_GETCONTIGUOUS(oarr);
```
Comment (dkurt) [inline modules/python/src2/cv2.cpp]: According to https://docs.scipy.org/doc/numpy-1.13.0/reference/c-api.array.html#c.PyArray_GETCONTIGUOUS it looks like `PyArray_GETCONTIGUOUS` should manage both cases.

File: modules/dnn/misc/python/test/test_dnn.py [hunk 4]
```diff
@@ -226,6 +226,67 @@ def test_async(self):
...
+                self.xstart = (inputShape[3] - targetShape[3]) // 2
+                self.yend = self.ystart + height
+                self.xend = self.xstart + width
+                return [[batchSize, numChannels, height, width]]
+            def forward(self, inputs):
+                return [inputs[0][:,:,self.ystart:self.yend,self.xstart:self.xend]]
+
+        cv.dnn_registerLayer('Crop', CropLayer)
+        proto = '''
+            name: "TestCrop"
```
Comment (alalek) [inline modules/dnn/misc/python/test/test_dnn.py]: Please remove unnecessary indentation:
```
            proto = '''\
name: "TestCrop"
...
}'''
```

File: modules/dnn/misc/python/test/test_dnn.py [hunk 5]
```diff
@@ -226,6 +226,67 @@ def test_async(self):
...
+                height, width = targetShape[2], targetShape[3]
+                self.ystart = (inputShape[2] - targetShape[2]) // 2
+                self.xstart = (inputShape[3] - targetShape[3]) // 2
+                self.yend = self.ystart + height
+                self.xend = self.xstart + width
+                return [[batchSize, numChannels, height, width]]
+            def forward(self, inputs):
+                return [inputs[0][:,:,self.ystart:self.yend,self.xstart:self.xend]]
+
+        cv.dnn_registerLayer('Crop', CropLayer)
```
Comment (alalek) [inline modules/dnn/misc/python/test/test_dnn.py]: What is happen in override attempt of existed layer?

Comment (dkurt) [reply modules/dnn/misc/python/test/test_dnn.py]: There is a stack of registered layers. So if there was a registered layer, new version will be used until dnn_unregister call. BTW, @l-bat, I think we should unregister it after the test similar to this sample - https://github.com/opencv/opencv/blob/master/modules/dnn/test/test_layers.cpp#L1303

File: modules/dnn/misc/python/test/test_dnn.py [hunk 6]
```diff
@@ -226,6 +226,73 @@ def test_async(self):
...
+        '}\n' \
+        'layer {\n' \
+        '  name: "Crop"\n' \
+        '  type: "CropCaffe"\n' \
+        '  bottom: "input"\n' \
+        '  bottom: "roi"\n' \
+        '  top: "Crop"\n' \
+        '} '
+
+        net = cv.dnn.readNetFromCaffe(str.encode(proto))
```
Comment (alalek) [inline modules/dnn/misc/python/test/test_dnn.py]: There are several functions with the same name. Code should call the right overload.
Try to wrap input into `bytearray(proto.encode())` (just string is a file name).

P.S. Revert multiline string back.

**Pull request merged**